#  Pre-lecture homework Answer of question1

In [3]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

# Question 2

In [29]:
df.shape

(391, 11)

observations: These are the individual entries or rows in the DataFrame. Each observation typically represents a single data point, entity, or case, and it is indexed by the rows. In the DataFrame I am using, each observation would correspond to a row that potentially represents a villager, assuming the data is about individuals.

Variables: These are the different attributes or features that describe each observation, and they are indexed by the columns in my DataFrame. Each variable represents a type of information that is collected about the observations. For example, in the case above, variables might include attributes like name, gender, species, etc.

# Question3

In [26]:
df.describe()

row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

In [6]:
df['gender'].value_counts()

gender
male      204
female    187
Name: count, dtype: int64

In [7]:
df['personality'].value_counts()

personality
lazy      60
normal    59
cranky    55
snooty    55
jock      55
peppy     49
smug      34
uchi      24
Name: count, dtype: int64

In [8]:
df['birthday'].value_counts()

birthday
1-27     2
12-5     2
7-31     2
3-26     2
8-3      2
        ..
4-3      1
10-26    1
7-23     1
12-8     1
3-8      1
Name: count, Length: 361, dtype: int64

# Question 4

In [5]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

In [14]:
df.shape

(891, 15)

In [24]:
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

# Question 5

df.shape: Purpose: df.shape is an attribute of the DataFrame that returns a tuple reflecting its dimensions, specifically the number of rows and columns it contains. Output: The output is a tuple where the first element represents the number of rows and the second represents the number of columns, formatted as (number_of_rows, number_of_columns). Use Case: Employ df.shape for a quick assessment of the size of my dataset, including the total number of entries and the number of attributes each entry has.

df.describe(): Purpose: This function generates summary statistics relevant to the DataFrame columns. For numeric data, it yields metrics like mean, standard deviation, and quartiles, and it can also summarize categorical data to show unique counts, the most common category, and its frequency. Output: The output is a DataFrame where each row corresponds to a different statistic and each column corresponds to a numerical (or specified categorical) column from the original DataFrame. Use Case: Utilize df.describe() to examine the statistical properties of numeric data and the composition of categorical variables within my dataset, providing insights into the distribution, variation, and central tendencies of your data.

# Question 6

count: The number of non-null entries in the column. This indicates how many data points are available for that variable, which is important for assessing sample size and missing data.

Mean (Average Value): The average of all non-null values in the column. It is calculated by summing up all the values and dividing by the count of values. The mean provides a measure of the central tendency of the data.

Standard Deviation: A measure of the amount of variation or dispersion in a set of values. A low standard deviation means that the values tend to be close to the mean, while a high standard deviation means that the values are spread out over a wider range.

Min (Minimum): The smallest value in the column. This gives an idea of the lower bound of the data range.

First Quartile (25%): Also known as the lower quartile, it is the median of the data values below the median in a dataset. Essentially, 25% of the data points are less than or equal to this value.

Median (50%): The middle value of the dataset when it is ordered from least to greatest. If the dataset has an even number of observations, the median is the average of the two middle numbers. This is a better measure of central tendency when the data is skewed because it is less affected by outliers and extreme values.

Third Quartile (75%): Also known as the upper quartile, it is the median of the data values above the median in a dataset. This means that 75% of the data points are less than or equal to this value.

Max (Maximum): The largest value in the column. This gives an idea of the upper bound of the data range.

# Question 7

Missing "Across Rows"
When data is missing "across rows," it means that certain observations (rows) have missing values in one or more columns. This pattern might indicate specific instances where data collection failed or where certain conditions led to incomplete data entries.

Handling Method: df.dropna():
This method is used to drop rows that contain any or all missing values, depending on the parameters set. The default behavior (df.dropna()) removes any row containing at least one missing value.
Use when: You want to ensure that analyses or models are only built with completely observed cases, or when the number of rows with missing data is small enough that their removal doesn't significantly reduce the dataset's size or representativeness.
Missing "Down Columns"
When data is missing "down columns," it indicates that an entire feature (column) has missing values, either completely or to a significant extent. This might happen if a particular measurement was difficult to obtain or irrelevant for a subset of observations.

Handling Method: del df['col'] or df.drop('col', axis=1):
These methods are used to remove a specific column from the DataFrame. The del statement removes the column in-place, while df.drop() can be configured to operate in-place or return a modified copy of the DataFrame.
Use when: A significant portion of a column is missing, which could bias any analysis including it, or when the column isn't relevant to your analysis/questions.
Deciding Between the Two
The choice between dropping rows or columns generally comes down to the nature of the missing data and the goals of your analysis:

Drop rows if the loss of data points doesn't significantly impact the dataset's integrity or if complete cases are necessary for a particular analysis.
Drop columns if the missing data is extensive within a column and imputation isn't feasible or would introduce significant bias, or if the column isn't crucial for your analysis.

#example of a "use case" in which using df.dropna() might be peferred over using del df['col']

#example of using df.dropna() might be peferred over using del df['col']
scenario: When having a data set of customer transactions where some transactions are missing the purchase_amount but have other important details like customer_id and transaction_date.

Reason to Use df.dropna(): If you want to analyze the total revenue and only care about transactions where the purchase_amount is available, you might prefer to drop rows with missing purchase_amount values. This is because keeping rows with missing purchase_amount would result in incomplete analysis or inaccurate results.

The code is written as
cleaned_df = df.dropna(subset=['purchase_amount'])

#example of using del df['col'] might be preferred over using df.dropna()
scenario: When having a dataset of employee records, but the temporary_address column has a lot of missing data and is not essential for analysis focusing on permanent addresses and employment history.

Reason to Use del df['col']: If the column is largely empty and irrelevant for your analysis, it might be more efficient to remove the column entirely. This avoids the need to handle missing values within the column and simplifies the dataset.

The code is written as
del df['temporary_address']

#example of why applying del df['col'] before df.dropna() when both are used together could be important
importance: Applying del df['col'] before df.dropna() can be important if a column contains a high proportion of missing data and is not relevant for analysis. Removing such columns first reduces the complexity of the dataset and avoids unnecessary operations.For example, if having a column with 90% missing data, dropping rows based on that column might lead to losing a significant portion of dataset. By removing the column first, focusing  df.dropna() on the more relevant columns and retain a larger portion of your dataset.

The code is written as
del df['irrelevant_column']

Then drop rows with missing values in the remaining columns
cleaned_df = df.dropna()

#example of Removing Missing Data and Justification
dataset:assuming a dataset with columns ['A', 'B', 'C', 'D'], where column A has a lot of missing values, and columns B, C, and D are critical for the analysis.

before report
the code is
print(df.isna().sum())

assuming the output is
A    100
B     10
C      5
D      0

approach:Remove Column A: Since it has too many missing values and is not essential for the analysis.Drop Remaining Rows with Missing Values: For columns B, C, and D.

implecation is:# Remove the column with excessive missing values
del df['A']
cleaned_df = df.dropna()
print(cleaned_df.isna().sum())

after report:
print(cleaned_df.isna().sum())
assuming the output is
B    0
C    0
D    0

justification: Removing column A ensures that we do not work with data that has too many missing entries, which might skew analysis or increase computational complexity. After removing irrelevant columns, dropping rows with missing values in the critical columns ensures that the data used in analysis is complete and reliable, thus maintaining data integrity and improving the quality of insights derived from the analysis.





In [25]:
del df['col']

KeyError: 'col'

# It's wrong because the name 'col' doesn't exist

#Summary of Exchanges
Introduction to df.describe() and df.shape:
df.describe(): Provides summary statistics for numerical columns in a DataFrame, including count, mean, standard deviation, min, quartiles, and max values.
df.shape: Returns the dimensions of the DataFrame, indicating the number of rows and columns.
Examples of Using df.describe() and df['column'].value_counts():
df.describe(): Example usage for obtaining summary statistics for numerical columns.
df['column'].value_counts(): Example usage for counting unique values in a categorical column.
Handling Missing Data:
df.dropna(): Used to remove rows with missing values, useful when you need to analyze data with complete entries.
del df['col']: Used to remove columns with excessive missing values or irrelevant columns, simplifying the dataset.
Use Cases:
Use Case for df.dropna(): Preferred when you need to retain important columns but discard rows with missing values in those columns (e.g., analyzing transactions with available amounts).
Use Case for del df['col']: Preferred when a column has too many missing values and is not relevant for analysis (e.g., removing a column with many missing temporary addresses).
Importance of Order:
Applying del df['col'] before df.dropna() is important if a column with many missing values is irrelevant, to avoid unnecessary operations and retain more useful data.
Removing Missing Data:
Approach: Remove irrelevant columns with excessive missing values first, then apply df.dropna() to clean the remaining data.
Example: Removing column A due to excessive missing values and then dropping rows with missing values in other columns (B, C, D).
Error Handling:
Issue: Encountered a KeyError when trying to delete a non-existent column.
Resolution: Check existing columns with df.columns, ensure the column name exists, and use the correct name for deletion.

# Question 8

In [26]:
df.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

In [27]:
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [17]:
df.groupby("pclass")["age"].describe()

count       mean        std   min   25%   50%   75%   max
pclass                                                           
1       186.0  38.233441  14.802856  0.92  27.0  37.0  49.0  80.0
2       173.0  29.877630  14.001077  0.67  23.0  29.0  36.0  70.0
3       355.0  25.140620  12.495398  0.42  18.0  24.0  32.0  74.0

In [19]:
#a different example from the "titanic" data set
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
grouped_describe_fare = df.groupby("embarked")["fare"].describe()
print(grouped_describe_fare)



          count       mean        std     min       25%    50%        75%  \
embarked                                                                    
C         168.0  59.954144  83.912994  4.0125  13.69795  29.70  78.500025   
Q          77.0  13.276030  14.188047  6.7500   7.75000   7.75  15.500000   
S         644.0  27.079812  35.887993  0.0000   8.05000  13.00  27.900000   

               max  
embarked            
C         512.3292  
Q          90.0000  
S         263.0000  


#Assuming you've not yet removed missing values in the manner of question "7" above, df.describe() would have different values in the count value for different data columns depending on the missingness present in the original data. Why do these capture something fundamentally different from the values in the count that result from doing something like df.groupby("col1")["col2"].describe()?

Answer:The df.describe() and df.groupby("col1")["col2"].describe() methods offer distinct types of summaries and insights about a DataFrame, especially when addressing missing values.

Key Differences

df.describe():
Purpose: Provides statistical summaries for numerical columns across the entire DataFrame.
Count in df.describe(): Shows the number of non-missing (non-NaN) entries in each numerical column. Columns with missing values will have a count lower than the DataFrame's total number of rows. This measure does not apply to categorical columns.
Insight: Offers overall descriptive statistics (mean, standard deviation, minimum, maximum, quartiles) for each column, useful for gauging the general data distribution. However, it does not consider variations in data across different groups within the DataFrame.

df.groupby("col1")["col2"].describe():
Purpose: Gives a statistical summary for a specific column (col2) within each subgroup defined by another column (col1).
Count in df.groupby("col1")["col2"].describe(): Indicates the number of non-missing entries in col2 for each col1 subgroup. The analysis is conducted separately for each group, and the count is specific to that group.
Insight: Aids in understanding how the distribution of values in a column varies among groups. It reveals if certain groups have more missing values or exhibit different statistical distributions.

Fundamental Differences
Scope of Analysis:
df.describe() provides a comprehensive overview of the data, summarizing the values of columns across the entire DataFrame.
df.groupby("col1")["col2"].describe() offers a detailed view, summarizing the values of a column within specific groups determined by another column.
Impact of Missing Values:
In df.describe(), the count reflects the total present values in each column, not considering how missing values are distributed among groups.
In df.groupby("col1")["col2"].describe(), the count shows the non-missing values within each group, which may differ based on the completeness of data in each group.

#example
#Without Removing Missing Values,the code is
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
print(df['age'].describe())
print(df.groupby("pclass")["age"].describe())

#Output for df['age'].describe():

count    714.000000
mean      29.699118
std       14.526497
min        0.420000
25%       20.125000
50%       28.000000
75%       38.000000
max       80.000000
Name: age, dtype: float64

#count: 714 non-missing values out of 891 total rows.

#Output for df.groupby("pclass")["age"].describe():
            count        mean         std   min    25%   50%    75%   max
pclass                                                                   
1        216.000000  38.233440  14.235776  22.0  30.000  35.0  44.000  80.0
2        184.000000  29.877630  14.256279  23.0  22.000  27.0  36.000  70.0
3        491.000000  24.734245  15.014139  0.42  17.000  22.0  30.000  80.0
#Counts for groups:
Class 1: 216 non-missing values.
Class 2: 184 non-missing values.
Class 3: 491 non-missing values.


#I showed that i report the error to chatgpt in the following link

https://chatgpt.com/share/521a5c8a-b1bc-4117-9fd6-367db5f7ed9e

# Question 9
Yes